File: test.spritebuilder/Source/TDEngine/Unit/Tower/Shell.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package Unit.Tower;

import Engine.PathEngine;
import Engine.SolderEngine;
import Unit.Solder.Solder;
import Utils.Point2D;
import Utils.StackObject;
import internalObject.ShellInternal;
 
 
 
import java.util.List;

/**
 *
 * @author Константин
 */
public class Shell extends StackObject implements ShellInternal{
      public Object userData;
    double speedShell = 8;
    int distanceToExploid = 7;
    Point2D  currentPlace;
    double attackStrong;
    Solder  target;
    public void init(double attackStrong, Solder attackTarget,Point2D currentPlace,double speedShell,int distanceToExploid) {
       this.attackStrong = attackStrong;
       this.target = attackTarget;
       this.currentPlace = currentPlace;
       this.speedShell = speedShell;
       this.distanceToExploid = distanceToExploid; 
    }

    public void DoStep(PathEngine lp,SolderEngine sold) {
     
        if( !exploided )
        { 
         Point2D pointToNext = target.getCurrentState();
        if(pointToNext == null)
        {    exploid(sold);
             return;
        }
        
        if(this.currentPlace.distance(pointToNext) < distanceToExploid) {
            exploid(sold);
            return;
        }
        
        this.currentPlace =   lp.calcNextDistanceForShell( this.currentPlace,
              target.getCurrentState(),speedShell);
           if(this.currentPlace == null)
                  exploid(sold);
        }
      
    }
      boolean exploided = false;
      public boolean isExploided(){return exploided;}
    private void exploid(SolderEngine sold) {
        exploided = true;
      List<Solder> targets =  sold.getSolderNearPlace(this.currentPlace,distanceToExploid);
        for (Solder solder : targets) {
            solder.Hit(attackStrong);
        }
    }

    
    @Override
    public Point2D getCurrent() {
        return this.currentPlace;
    }

    @Override
    public Point2D getTarget() {
     return target.getCurrentState();
    }

    @Override
    public void clearObject() {
       exploided = false;
    }

  

    @Override
    public boolean getIsBlowedOut() {
     return this.isExploided();
    }

  
  

    
}

File: TowerDefence.spritebuilder/Unit/Tower/Shell.swift
//
//  Shell.swift
//  TowerDefence
//
//  Created by koveres on 07.12.14.
//  Copyright (c) 2014 Apportable. All rights reserved.
//

import Foundation



public class Shell : StackObject , ShellInternal{
     var speedShell:CDouble = 8;
    var distanceToExploid:CInt = 7;
    var currentPlace:CGPoint?;
    var attackStrong:CDouble;
    var  target:Solder;
    
    
    
    public  init( attackStrong1:CDouble,  attackTarget:Solder, currentPlace1:CGPoint,  speedShell1:CDouble, distanceToExploid1:CInt,grData1:AnyObject) {
        
         attackStrong = attackStrong1;
         target = attackTarget;
         currentPlace = currentPlace1;
         speedShell = speedShell1;
         distanceToExploid = distanceToExploid1;
        
        super.init(data: grData1);
        
    }
    
    public func DoStep( lp:PathEngine, sold:SolderEngine) {
        
        if( !exploided )
        {
            var pointToNext:CGPoint? = target.getCurrentState();
            if(pointToNext == nil)
            {    exploid(sold);
                return;
            }
            
            if(ccpDistance(currentPlace!, pointToNext!) < CGFloat(distanceToExploid)) {
                exploid(sold);
                return;
            }
            
            currentPlace =   lp.calcNextDistanceForShell( currentPlace!,
                targetPoint: target.getCurrentState()!,speedShell: speedShell);
            if( currentPlace == nil){
                exploid(sold);}
        }
        
    }
    var exploided:Bool = false;
    public func isExploided() -> (Bool){return exploided;}
    private  func exploid( sold:SolderEngine) {
        exploided = true;
        var targets:[Solder] =  sold.getSolderNearPlace( currentPlace!,distanceToExploid: distanceToExploid);
        for  solder in targets  {
            solder.Hit(attackStrong);
        }
    }
    
    
 
    public func getCurrent() -> (CGPoint?) {
        return  currentPlace;
    }
    
   
    public func getTarget() -> (CGPoint?) {
        return target.getCurrentState();
    }
    
 
    public override func clearObject()  {
        exploided = false;
    }
    
    
    
 
    public func getIsBlowedOut() -> (Bool) {
        return isExploided();
    }
    
    
    
    
    
}
